ethcore/light/src/net/mod.rs

```diff
@@ -0,0 +1,506 @@
+		let mut peers = self.peers.write();
+
+		let peer_info = match peers.get_mut(peer) {
+			Some(info) => info,
+			None => return Ok(()),
+		};
+
+		// update status.
+		{
+			// TODO: punish peer if they've moved backwards.
+			let status = &mut peer_info.status;
+			let last_head = status.head_hash;
+			status.head_hash = announcement.head_hash;
+			status.head_td = announcement.head_td;
+			status.head_num = announcement.head_num;
+			status.last_head = Some((last_head, announcement.reorg_depth));
+		}
+
+		// update capabilities.
+		{
+			let caps = &mut peer_info.capabilities;
+			caps.serve_headers = caps.serve_headers || announcement.serve_headers;
+			caps.serve_state_since = caps.serve_state_since.or(announcement.serve_state_since);
+			caps.serve_chain_since = caps.serve_chain_since.or(announcement.serve_chain_since);
+			caps.tx_relay = caps.tx_relay || announcement.tx_relay;
+		}
+
+		// TODO: notify listeners if new best block.
+
+		Ok(())
+	}
+
+	// Handle a request for block headers.
+	fn get_block_headers(&self, peer: &PeerId, io: &NetworkContext, data: UntrustedRlp) -> Result<(), Error> {
+		const MAX_HEADERS: usize = 512;
+
+		let mut present_buffer = match self.peers.read().get(peer) {
+			Some(peer) => peer.local_buffer.clone(),
+			None => {
+				debug!(target: "les", "Ignoring announcement from unknown peer");
+				return Ok(())
+			}
+		};
+
+		self.flow_params.recharge(&mut present_buffer);
+		let req_id: u64 = try!(data.val_at(0));
+
+		let req = request::Headers {
+			block: {
+				let rlp = try!(data.at(1));
+				(try!(rlp.val_at(0)), try!(rlp.val_at(1)))
+			},
+			max: ::std::cmp::min(MAX_HEADERS, try!(data.val_at(2))),
+			skip: try!(data.val_at(3)),
+			reverse: try!(data.val_at(4)),
+		};
+
+		let max_cost = self.flow_params.compute_cost(request::Kind::Headers, req.max);
+		try!(present_buffer.deduct_cost(max_cost));
+
+		let response = self.provider.block_headers(req);
+		let actual_cost = self.flow_params.compute_cost(request::Kind::Headers, response.len());
+
+		let cur_buffer = match self.peers.write().get_mut(peer) {
+			Some(peer) => {
+				self.flow_params.recharge(&mut peer.local_buffer);
+				try!(peer.local_buffer.deduct_cost(actual_cost));
+				peer.local_buffer.current()
+			}
+			None => {
+				debug!(target: "les", "peer disconnected during serving of request.");
+				return Ok(())
+			}
+		};
+
+		io.respond(packet::BLOCK_HEADERS, {
+			let mut stream = RlpStream::new_list(response.len() + 2);
+			stream.append(&req_id).append(&cur_buffer);
+
+			for header in response {
+				stream.append_raw(&header, 1);
+			}
+
+			stream.out()
+		}).map_err(Into::into)
+	}
+
+	// Receive a response for block headers.
+	fn block_headers(&self, _: &PeerId, _: &NetworkContext, _: UntrustedRlp) -> Result<(), Error> {
+		unimplemented!()
+	}
+
+	// Handle a request for block bodies.
+	fn get_block_bodies(&self, peer: &PeerId, io: &NetworkContext, data: UntrustedRlp) -> Result<(), Error> {
```

Comment:
lots of duplicated code here from `get_block_headers` -- I stopped implementing request handlers here for the moment as I realized there is probably a decent opportunity for abstraction.

These parts are common to basically every request handler:
```
  - check if peer exists
  - deserialize request UntrustedRlp -> (req_id, T)
  - check cost: &T -> U256
  - request from provider: T -> U
  - actual cost: &U -> U256
  - response: (req_id, U) -> (packet, Vec<u8>)
```

So it ought to be possible to build some kind of request-handler builder with a fluent API that handles all of the common parts.